### PR TITLE
Allow to specify image by digest. Fixes #1670

### LINKED ISSUE #1670: Add support for referring to images by digest in docker-compose.yml
Pulling by image digest is supported since docker 1.6.0 (https://github.com/docker/docker/pull/11109)

```
docker pull busybox@sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d
```

However putting an image digest for a service in `docker-compose.yml` leads to

```
Pulling busybox (busybox@sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d)...
Invalid repository name (busybox@sha256), only [a-z0-9-_.] are allowed
```

Sample `docker-compose.yml`

```
service:
  image: busybox@sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d
```

```
docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): darwin/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```

```
docker-compose --version
docker-compose version: 1.3.1
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1j 15 Oct 2014
```

### PR DESCRIPTION
**Pull request by mrfuxi**
Signed-off-by: Karol Duleba <mr.fuxi@gmail.com>

### CODE REVIEW

**Comment by aanand** [comment]
Thanks! LGTM

File: compose/service.py [hunk 1]
```diff
@@ -780,14 +780,22 @@ def build_container_name(project, service, number, one_off=False):
...
+    if digest_sep in s:
+        separator = digest_sep
+        repo, tag = s.rsplit(digest_sep, 1)
+
+    elif tag_sep in s:
+        repo, tag = s.rsplit(tag_sep, 1)
+        if "/" in tag:
+            repo, tag = s, ""
+
+    return repo, tag, separator
```
Comment (mnowster) [inline compose/service.py]: I'm finding this function hard to parse/has a cognitive load to understand intended behaviour. To help me understand, I re-wrote it.:

`def parse_repository_tag(repo_path):
    tag_separator = ":"
    digest_separator = "@"

    if digest_separator in repo_path:
        repo, tag = repo_path.rsplit(digest_separator, 1)
        return repo, tag, digest_separator

    if tag_separator in repo_path:
        repo, tag = repo_path.rsplit(tag_separator, 1)
        if "/" in tag:
            return repo_path, "", tag_separator

        return repo, tag, tag_separator

    return repo_path, "", tag_separator
`

How do you feel about this way of expressing the function? Using fully expanded variable names really helps with readability.

Comment (mnowster) [inline compose/service.py]: A doc string for the function might be helpful too.

**Comment by mnowster** [comment]
It's really nice to review PRs that already include tests and updates to docs too, thoughtful good work :balloon:

File: compose/service.py (hunk 1 above)
Comment (mrfuxi) [inline compose/service.py]: Doc string is definitely worth adding.
I like a explicit and easy to understand variable names as well as early returns.
However 4 returns (3 in the row) in 16 lines of code is a bit to much. How do you think? @mnowster

Comment (mrfuxi) [inline compose/service.py]:
```python
def parse_repository_tag(repo_path):
    tag_separator = ":"
    digest_separator = "@"

    if digest_separator in repo_path:
        repo, tag = repo_path.rsplit(digest_separator, 1)
        return repo, tag, digest_separator

    repo, tag = repo_path, ""
    if tag_separator in repo_path:
        repo, tag = repo_path.rsplit(tag_separator, 1)
        if "/" in tag:
            repo, tag = repo_path, ""

    return repo, tag, tag_separator
```

Comment (mnowster) [inline compose/service.py]: :+1: yes, I like this solution, good improvement.

If you update a commit with this approach and include a docstring, then I'm happy to merge. Thanks!

**Comment by mrfuxi** [comment]
Function updated as discussed

**Comment by mnowster** [comment]
LGTM
